src/System.Private.CoreLib/shared/System/Number.Formatting.cs

```diff
@@ -1095,6 +1097,13 @@ private static unsafe void UInt32ToNumber(uint value, ref NumberBuffer number)
         private static unsafe string UInt32ToDecStr(uint value, int digits)
         {
             int bufferLength = Math.Max(digits, FormattingHelpers.CountDigits(value));
+
+            // For single-digit values that are very common, especially 0 and 1, just return cached strings.
+            if (bufferLength == 1)
+            {
+                return s_singleDigitStringCache[value];
```

Comment:
```
if (value <= 10)
{
     return s_smallNumberStringCache[value];
}
```

Reply:
That also needs to factor in digits.  This can be called with something like digits == 5 and value == 2, and it needs to return "00002".

Reply:
In that case, should we be validating that digits is -1/etc, or is it safe to always return single digits as "digit" if bufferLength is 1?

Reply:
> In that case, should we be validating that digits is -1/etc., or is it safe to always return single digits as "digit" if bufferLength is 1?

No matter what the `digits` value is, there will always be at least one digit rendered.